### PR TITLE
568 Enable searching for nodes

### LINKED ISSUE #568: Search view to also allow searching for links and nodes. make similar toggle functionality for filtering as current transit types

### PR DESCRIPTION
**Pull request by kronken**
Closes #568

### CODE REVIEW

**Comment by Divinare** [comment]
https://github.com/HSLdevcom/jore-map-ui/pull/586

**Comment by Divinare** [comment]
What if toggling off "Linjat ja Reitit" would disable transitTypeToggle buttons?

**Comment by Divinare** [comment]
Testing:

If "Linjat ja Reitit" is toggled off, node results can't be scrolled down.

File: src/components/shared/searchView/SearchResults.tsx [hunk 1]
```diff
@@ -142,6 +146,16 @@ class SearchResults extends React.Component<ISearchResultsProps, ISearchResultsS
...
 
+    private isLine(item: INodeBase | ILine): item is ILine {
+        return (
+            (item as ILine).routes !== undefined ||
+            (item as ILine).transitType !== undefined
+        ) && (
+            (item as INodeBase).shortId === undefined ||
+            (item as INodeBase).type === undefined
+        );
+    }
```
[the remark this answers is not in the document]
Comment (Divinare) [reply src/components/shared/searchView/SearchResults.tsx]: If not possible with any cleaner way, skip this

File: src/stores/searchResultStore.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        });
+    }
+
+    public getFilteredItems =
+        (
+            searchInput: string,
+            transitTypes: TransitType[],
+            includeLines: boolean,
+            includeNodes: boolean,
+            ): (INodeBase | ILine)[] => {
```
Comment (Divinare) [inline src/stores/searchResultStore.ts]: Use import searchStore (look at geometryEventStore as an example) so you don't have to pass any parameters here

File: src/stores/searchResultStore.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+            // Filter by route.name
+            return line.routes
+                .map(route => route.name.toLowerCase())
+                .some(name => name.indexOf(searchInput) > -1);
+        });
+    }
+
+    private getFilteredNodes = (searchInput: string) => {
+        return this._allNodes.filter((node) => {
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/stores/searchResultStore.ts]: I usually prefer leaving typings out when setting arrow functions as parameters to array methods. When type of the parameters is clear, such as in this case, I find extra typings just to add noise to the code.

File: src/stores/searchResultStore.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        this._allLines = lines;
+    }
+
+    @action
+    public setAllNodes = (nodes: INodeBase[]) => {
+        this._allNodes = nodes;
+    }
+
+    private getFilteredLines = (searchInput: string, transitTypes: TransitType[]) => {
+        return this._allLines.filter((line) => {
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/stores/searchResultStore.ts]: Same thing

File: src/components/shared/searchView/SearchResults.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (kronken) [reply src/components/shared/searchView/SearchResults.tsx]: This is the documented way for type checking:
https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards

**Pull request merged**